File: Solutions/394_Decode_String/394_Decode_String.js
//https://leetcode.com/problems/decode-string/discuss/245538/Javascript-using-Regex.
/**
 * @param {string} s
 * @return {string}
 */
function decodeString1(s) {
    return s.replace(/(\d+\[\w+\])/gi, function(item) {
        var match = /(\d+)\[(\w+)\]/.exec(item);
        var repeat = parseInt(match[1]);
        var pattern = match[2];
        
        var result = "";
        while(repeat-- > 0) {
            result += pattern;
        }
        return result;
    });
};

var decodeString = function(s) {
    while(/\d+\[\w+\]/gi.test(s)) {
        s = decodeString1(s);
    }
    return s;
};

//Fourth fastest
var decodeString = function (s) {
    const ans = [];
  
    for (const char of s) {
      if (char === ']') {
        let cur = ans.pop();
        let str = '';
  
        while (cur !== '[') {
          str = cur + str;
          cur = ans.pop();
        }
  
        let num = '';
        cur = ans.pop();
        while (!isNaN(cur)) {
          num = cur + num;
          cur = ans.pop();
        }
  
        ans.push(cur);
        ans.push(str.repeat(Number(num)));
      } else {
        ans.push(char);
      }
    }
  
    return ans.join('');
  };

//Third fastest
/**
 * @param {string} s
 * @return {string}
 */

/*
The pattern could begin with a number K, followed by opening brace '[', followed by a string
    - Then it could be immediately closed by a ']'
    - Or, there could be another nested pattern     K[string K[string]] 

- If there is a nested pattern, we have to decode the innermost pattern first
    - Push each character into the stack until we meet a ']'
    - Once we meet a ']', we start decoding the pattern 
    
    
1) 2 Stacks, one for numbers, one for substring
2) If we see a number, push into numberStack
3) When we meet an opening bracket, 
    
    
    // approach: stack
    // loop thru string
    // if s[i] is a number, add that to the numStack
    // if "[", you know its the start of a current string of letters
    ///if "]", the current string ends and is repeated for numstack.pop() times
    // add above to a strStack
    
    
    s = "3[a]2[bc]"
    
    currNum = '3'
    
*/

var decodeString = function(s) {
    
    let numStack = []
    let stringStack = []
    let currString = ''
    let currNum = ''
    
    for (let i = 0; i < s.length; i++) {
        let element = s[i]
        
        // Is a number
        if (isNaN(element) === false) {
            currNum += element
        }
            
        // We know the number just ended. 
        else if (element === '[') {
            numStack.push(Number(currNum))
            currNum = ''
            stringStack.push(currString)
            currString = ''
        }
        
        else if (element === ']') {
            let count = numStack.pop()
            currString = stringStack.pop() + currString.repeat(count)
        }
        
        // We are dealing with a string
        else {
            currString += element
        }        
    }
    
    return currString
};

//Second fastest
var decodeString = function(s) {
    // base case ""
       //if (s.length == 0) return "";
      
       let index = 0, sb = '';
       while (index < s.length) {
           let c = s.charAt(index);
           if (isNaN(c)) {
               sb += c;
               index++;
           } else if (!isNaN(c)) {
               let end = index;
               while (!isNaN(s.charAt(end))) end++;
               let num = parseInt(s.substring(index, end));
               index = end;
               let open = 1;
               end++;
               while (end < s.length && open !== 0) {
                   if (s.charAt(end) == '[') open++;
                   else if (s.charAt(end) == ']') open--;
                   end++;
               }
               let encoded = decodeString(s.substring(index+1, end-1));
               for (let i = 0; i < num; i++) sb += encoded;
               index = end;
           }
       }
   //console.log(sb)
       return sb;
  
};

decodeString("3[a2[c]]")


//Fastset
var decodeString = function(s) {
    // base case ""
       if (s.length == 0) return "";
      
       let index = 0, sb = '';
       while (index < s.length) {
           let c = s.charAt(index);
           if (isNaN(c)) {
               sb += c;
               index++;
           } else if (!isNaN(c)) {
               let end = index;
               while (!isNaN(s.charAt(end))) end++;
               let num = parseInt(s.substring(index, end));
               index = end;
               let open = 1;
               end++;
               while (end < s.length && open !== 0) {
                   if (s.charAt(end) == '[') open++;
                   else if (s.charAt(end) == ']') open--;
                   end++;
               }
               let encoded = decodeString(s.substring(index+1, end-1));
               for (let i = 0; i < num; i++) sb += encoded;
               index = end;
           }
       }
       return sb;
  
};

File: Solutions/394_Decode_String/394_Decode_String.rb
#Fourth fastest
def decode_string(s)
    loop do
      break unless s.match?(%r{((\d+)\[(([a-z]|[A-Z])*)\])})
      s.scan(%r{((\d+)\[(([a-z]|[A-Z])*)\])}).each do |subset|
        s.gsub!(subset[0], Array.new(subset[1].to_i, subset[2]).join)
      end
    end
  
    s
  end

#Third fastest
def decode_string(s)
    p s
    result = ''
    arr= []
    count = 0
    multi_letter = ''
    digit = ''
  
    s.each_char.with_index(1) do |letter, i |
      if letter == "[" && count == 0
        unless multi_letter == ''
          arr.push([multi_letter.slice(0,multi_letter.length),1])
        end
        multi_letter = ''
  
      elsif letter == "]" && count == 1
        arr.push [multi_letter,digit]
        digit = ''
        multi_letter = ''
      elsif count > 0 
          multi_letter.concat(letter)
      elsif count == 0 
        if  (letter =~ /\d/) == nil
  
          multi_letter.concat(letter)
        end
      end
      p arr
      if count == 0 && (letter =~ /\d/) == 0
        digit.concat letter
      end
      
      if letter == '['    
        count +=1
      elsif letter == ']'    
        count -=1
      end
    end
    
    unless multi_letter==''
      arr.push [multi_letter,1]
    end
  
    until arr.empty?
      add_string = arr.shift
      if add_string[0].include?('[')
        add_string[0] = decode_string(add_string[0])
      end
      result.concat(add_string[0]*add_string[1].to_i)
    end
     result
  end

#Second fastest
def decode_string(s)
    stack = []
    s.split('').each do |c|
        stack.push(c) and next unless c == ']'
        temp = ''
        ch = stack.pop
        while ch != '[' 
            temp = ch + temp
            ch = stack.pop
        end
        n = ''
        (n = stack.pop + n) while numeric?(stack[-1])
        temp = temp * n.to_i
        stack.push(temp)
    end
    return stack.join('')
end
def numeric? str
    Float(str) != nil rescue false
end

#Fastest
def decode_string(s)
    stack = []
    nums = "1234567890"
    
    (0...s.length).each do |i|
        if s[i] == "]"
            decoded = []
            while stack.last != "["
                decoded << stack.pop
            end
            stack.pop
            base = 1
            k = 0
            while !stack.empty? && nums.include?(stack.last)
                k = k + stack.pop.to_i * base
                base *= 10
            end
            
            while k != 0
                j = decoded.length - 1
                while j >= 0
                    stack.push(decoded[j])
                    j -= 1
                end
                k -= 1
            end
        else
            stack.push(s[i])
        end
    end
    stack.join("")
end